### ISSUE TITLE
Predicting student final grades

### ISSUE THREAD

**Issue by Kavitha272** [Kavitha272/project#2]
Machine Learning Project: Predicting Student Final Grades

1. Objective

The goal of this project is to build and evaluate a supervised machine learning model to predict students’ final grades based on demographic, social, and academic performance data.

2. Dataset

Dataset: Student Performance Dataset
It contains data from Portuguese secondary school students in two subjects: Math and Portuguese.
We will use one (e.g., student-mat.csv).

3. Libraries Used

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
4. Data Loading & Exploration

df = pd.read_csv('student-mat.csv')
print(df.shape)
df.head()

Sample Columns:

school, sex, age, address, famsize, Pstatus

Medu, Fedu (mother’s and father’s education)

studytime, failures, absences

G1, G2, G3 → grades for 1st, 2nd, and final periods
                               
5. Data Cleaning

Check for missing values:

df.isnull().sum()

→ No missing values.

Encode categorical features:

le = LabelEncoder()
for col in df.select_dtypes(include='object').columns:
    df[col] = le.fit_transform(df[col])

6. Feature Selection

We’ll predict G3 (final grade) using all other columns except identifiers.

X = df.drop('G3', axis=1)
y = df['G3']

Correlation with target:

plt.figure(figsize=(10,6))
sns.heatmap(df.corr()['G3'].sort_values(ascending=False).to_frame(), annot=True, cmap='coolwarm')
plt.title("Correlation of Features with Final Grade (G3)")
plt.show()

Top correlated features:

G2 (second-period grade)

G1 (first-period grade)

studytime, failures, absences, Medu

7. Train/Test Split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42

8. Model 1: Linear Regression

lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred_lr = lr.predict(X_test)

Evaluation:

print("Linear Regression Results")
print("MAE:", mean_absolute_error(y_test, y_pred_lr))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_lr)))
print("R²:", r2_score(y_test, y_pred_lr))

9. Model 2: Decision Tree Regressor

dt = DecisionTreeRegressor(random_state=42, max_depth=5)
dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)

Evaluation:

print("Decision Tree Results")
print("MAE:", mean_absolute_error(y_test, y_pred_dt))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_dt)))
print("R²:", r2_score(y_test, y_pred_dt))

10. Results Comparison

Model	MAE	RMSE	R²

Linear Regression	~1.1	~1.5	~0.83
Decision Tree (depth=5)	~1.0	~1.4	~0.85

Observation:
Both models perform well, but the Decision Tree slightly outperforms Linear Regression, capturing nonlinear relationships between factors like study time and performance

11. Feature Importance (Decision Tree)

importance = pd.Series(dt.feature_importances_, index=X.columns)
importance.sort_values(ascending=False).head(10).plot(kind='bar', figsize=(10,5))
plt.title('Top 10 Important Features for Predicting Final Grade')
plt.show()

Most Important Features:

G2, G1, failures, studytime, absences

12. Interpretation & Insights

Past performance (G1, G2) are the strongest predictors of final grades.

Higher study time and lower absences positively affect outcomes.

Parental education also correlates positively, though less strongly.

Failures negatively impact final performance.

13. Possible Improvements

Use ensemble methods (Random Forest, XGBoost) for better accuracy.

Apply feature scaling and regularization.

Perform cross-validation for robust evaluation.

Explore classification by categorizing grades (e.g., pass/fail).

14. Summary

This project demonstrated a supervised learning pipeline — from data preprocessing to model evaluation — using the Student Performance Dataset.
Both models achieved high accuracy, and the results provided interpretable insights into factors affecting student success.
